File: experiments/build123d-vs-ocjs/native/samples.cpp
// Frontier OCCT 8.0.0 native C++ workloads mirroring ocjs/samples.mjs and
// python/samples.py.
//
// Every sample that batches >1 boolean operation uses the canonical
// `BRepAlgoAPI_BuilderAlgo` multi-tool form (`SetArguments + SetTools + Build`)
// — the iterative `Op(prev, next)` chain is the previous-baseline anti-pattern
// (preserved historically in F13's per-engine 09b/09 + 10b/10 ratios). See
// F14 — Frontier benchmark sample (see experiments/build123d-vs-ocjs/README.md).
// Performance — is the canonical comparison and uses these samples directly.

#include "samples.hpp"

#include <stdexcept>

#include <BRep_Tool.hxx>
#include <BRepAlgoAPI_Cut.hxx>
#include <BRepAlgoAPI_Fuse.hxx>
#include <BRepBuilderAPI_MakeEdge.hxx>
#include <BRepBuilderAPI_MakeWire.hxx>
#include <BRepFilletAPI_MakeFillet.hxx>
#include <BRepMesh_IncrementalMesh.hxx>
#include <BRepOffsetAPI_MakeFilling.hxx>
#include <BRepOffsetAPI_MakePipeShell.hxx>
#include <BRepOffsetAPI_ThruSections.hxx>
#include <BRepPrimAPI_MakeBox.hxx>
#include <BRepPrimAPI_MakeCylinder.hxx>
#include <ChFi3d_FilletShape.hxx>
#include <GeomAbs_Shape.hxx>
#include <Geom_Circle.hxx>
#include <Message_ProgressRange.hxx>
#include <NCollection_List.hxx>
#include <Poly_MeshPurpose.hxx>
#include <Poly_Triangulation.hxx>
#include <TopAbs_ShapeEnum.hxx>
#include <TopExp_Explorer.hxx>
#include <TopLoc_Location.hxx>
#include <TopoDS.hxx>
#include <TopoDS_Edge.hxx>
#include <TopoDS_Face.hxx>
#include <TopoDS_Shape.hxx>
#include <TopoDS_Wire.hxx>
#include <gp_Ax2.hxx>
#include <gp_Dir.hxx>
#include <gp_Pnt.hxx>

namespace {

inline void require(bool ok, const char* what) {
  if (!ok) throw std::runtime_error(what);
}

// `TopTools_ListOfShape` is deprecated in OCCT 8.0 in favour of the canonical
// `NCollection_List<TopoDS_Shape>` template instantiation. We use the canonical
// form directly so the C++ code mirrors the JS surface (`oc.NCollection_List_TopoDS_Shape`).
using ListOfShape = NCollection_List<TopoDS_Shape>;

}  // namespace

void sample01_primitive_box() {
  BRepPrimAPI_MakeBox mk(10, 20, 30);
  TopoDS_Shape shape = mk.Shape();
  require(!shape.IsNull(), "null shape");
}

void sample02_primitive_cylinder() {
  BRepPrimAPI_MakeCylinder mk(5, 15);
  TopoDS_Shape shape = mk.Shape();
  require(!shape.IsNull(), "null shape");
}

void sample03_boolean_fuse() {
  BRepPrimAPI_MakeBox b1(10, 10, 10);
  BRepPrimAPI_MakeBox b2(5, 5, 5);
  TopoDS_Shape s1 = b1.Shape();
  TopoDS_Shape s2 = b2.Shape();
  Message_ProgressRange pr;
  BRepAlgoAPI_Fuse fuse(s1, s2, pr);
  Message_ProgressRange pr2;
  fuse.Build(pr2);
  TopoDS_Shape out = fuse.Shape();
  require(!out.IsNull(), "null fuse");
}

// Multi-tool BRepAlgoAPI_Cut: 1 BOPDS init over base + 25 cylinder tools
// instead of 25 separate inits over (current, tool). See F14.
void sample04_boolean_cut_grid() {
  BRepPrimAPI_MakeBox baseMaker(100, 100, 10);
  TopoDS_Shape base = baseMaker.Shape();
  ListOfShape args, tools;
  args.Append(base);
  for (int i = 0; i < 5; ++i) {
    for (int j = 0; j < 5; ++j) {
      gp_Pnt origin(10 + i * 20.0, 10 + j * 20.0, -5);
      gp_Ax2 ax(origin, gp_Dir(0, 0, 1));
      BRepPrimAPI_MakeCylinder cylMaker(ax, 2, 20);
      tools.Append(cylMaker.Shape());
    }
  }
  BRepAlgoAPI_Cut cut;
  cut.SetArguments(args);
  cut.SetTools(tools);
  Message_ProgressRange pr;
  cut.Build(pr);
  TopoDS_Shape out = cut.Shape();
  require(!out.IsNull(), "null cut grid");
}

void sample05_loft_thru_sections() {
  BRepOffsetAPI_ThruSections loft(true, false, 1e-6);
  struct P { double z, radius; };
  const P profiles[] = {{0, 10}, {15, 5}, {30, 8}};
  for (const auto& p : profiles) {
    gp_Pnt center(0, 0, p.z);
    gp_Dir dir(0, 0, 1);
    gp_Ax2 ax(center, dir);
    Handle(Geom_Circle) circle = new Geom_Circle(ax, p.radius);
    BRepBuilderAPI_MakeEdge em(circle);
    TopoDS_Edge e = em.Edge();
    BRepBuilderAPI_MakeWire wm(e);
    TopoDS_Wire w = wm.Wire();
    loft.AddWire(w);
  }
  loft.CheckCompatibility(false);
  TopoDS_Shape shape = loft.Shape();
  require(!shape.IsNull(), "null loft");
}

void sample06_pipe_shell_sweep() {
  gp_Pnt p1(0, 0, 0), p2(0, 0, 30);
  BRepBuilderAPI_MakeEdge spineEdge(p1, p2);
  TopoDS_Edge eSpine = spineEdge.Edge();
  BRepBuilderAPI_MakeWire spineWire(eSpine);

  gp_Pnt axOrigin(0, 0, 0);
  gp_Dir axDir(0, 0, 1);
  gp_Ax2 ax(axOrigin, axDir);
  Handle(Geom_Circle) circle = new Geom_Circle(ax, 5);
  BRepBuilderAPI_MakeEdge profileEdge(circle);
  TopoDS_Edge eProf = profileEdge.Edge();
  BRepBuilderAPI_MakeWire profileWire(eProf);

  TopoDS_Wire sw = spineWire.Wire();
  BRepOffsetAPI_MakePipeShell pipeShell(sw);
  TopoDS_Wire pw = profileWire.Wire();
  pipeShell.Add(pw, false, false);
  Message_ProgressRange progress;
  pipeShell.Build(progress);
  pipeShell.MakeSolid();
  TopoDS_Shape shape = pipeShell.Shape();
  require(!shape.IsNull(), "null pipe");
}

void sample07_surface_filling_patch() {
  gp_Pnt p1(0, 0, 0), p2(10, 0, 0), p3(10, 10, 0), p4(0, 10, 0);
  BRepBuilderAPI_MakeEdge em1(p1, p2);
  BRepBuilderAPI_MakeEdge em2(p2, p3);
  BRepBuilderAPI_MakeEdge em3(p3, p4);
  BRepBuilderAPI_MakeEdge em4(p4, p1);
  TopoDS_Edge e1 = em1.Edge();
  TopoDS_Edge e2 = em2.Edge();
  TopoDS_Edge e3 = em3.Edge();
  TopoDS_Edge e4 = em4.Edge();

  BRepOffsetAPI_MakeFilling filling(3, 15, 2, false, 1e-3, 1e-4, 1e-1, 0.1, 8, 9);
  filling.Add(e1, GeomAbs_C0, true);
  filling.Add(e2, GeomAbs_C0, true);
  filling.Add(e3, GeomAbs_C0, true);
  filling.Add(e4, GeomAbs_C0, true);
  Message_ProgressRange pr;
  filling.Build(pr);
  TopoDS_Shape face = filling.Shape();
  require(!face.IsNull(), "null fill");
}

void sample08_fillet_all_edges() {
  BRepPrimAPI_MakeBox box(20, 20, 20);
  TopoDS_Shape boxShape = box.Shape();
  BRepFilletAPI_MakeFillet fillet(boxShape, ChFi3d_Rational);
  TopExp_Explorer explorer(boxShape, TopAbs_EDGE, TopAbs_SHAPE);
  while (explorer.More()) {
    const TopoDS_Shape& cur = explorer.Current();
    TopoDS_Edge edge = TopoDS::Edge(cur);
    fillet.Add(3, edge);
    explorer.Next();
  }
  Message_ProgressRange pr;
  fillet.Build(pr);
  TopoDS_Shape shape = fillet.Shape();
  require(!shape.IsNull(), "null fillet");
}

namespace {

// Multi-tool BRepAlgoAPI_Fuse: 1 BOPDS init over args + 39 box tools instead
// of 39 separate (prev, next) inits in a chain. See F13/F14.
TopoDS_Shape fuse_overlapping_boxes_multi_tool(int count, double spacing, double side) {
  ListOfShape args, tools;
  for (int i = 0; i < count; ++i) {
    gp_Pnt origin(i * spacing, 0, 0);
    BRepPrimAPI_MakeBox mk(origin, side, side, side);
    if (i == 0) args.Append(mk.Shape());
    else tools.Append(mk.Shape());
  }
  BRepAlgoAPI_Fuse fuse;
  fuse.SetArguments(args);
  fuse.SetTools(tools);
  Message_ProgressRange pr;
  fuse.Build(pr);
  return fuse.Shape();
}

}  // namespace

void sample09_fuse_many_boxes() {
  TopoDS_Shape shape = fuse_overlapping_boxes_multi_tool(40, 3.0, 4.0);
  require(!shape.IsNull(), "null multi-tool fuse");
}

void sample10_mesh_incremental() {
  TopoDS_Shape shape = fuse_overlapping_boxes_multi_tool(40, 3.0, 4.0);
  require(!shape.IsNull(), "null multi-tool fuse");
  BRepMesh_IncrementalMesh mesh(shape, 0.25, false, 0.5, false);
  (void)mesh;
  TopLoc_Location loc;
  TopExp_Explorer expf(shape, TopAbs_FACE, TopAbs_SHAPE);
  require(expf.More(), "no faces");
  const TopoDS_Shape& fc = expf.Current();
  TopoDS_Face face = TopoDS::Face(fc);
  Handle(Poly_Triangulation) tri = BRep_Tool::Triangulation(face, loc, Poly_MeshPurpose_NONE);
  require(!tri.IsNull() && tri->NbTriangles() >= 1, "no triangles");
}

const std::vector<Sample>& all_samples() {
  static const std::vector<Sample> kSamples = {
      {"01_primitive_box", sample01_primitive_box},
      {"02_primitive_cylinder", sample02_primitive_cylinder},
      {"03_boolean_fuse", sample03_boolean_fuse},
      {"04_boolean_cut_grid", sample04_boolean_cut_grid},
      {"05_loft_thru_sections", sample05_loft_thru_sections},
      {"06_pipe_shell_sweep", sample06_pipe_shell_sweep},
      {"07_surface_filling_patch", sample07_surface_filling_patch},
      {"08_fillet_all_edges", sample08_fillet_all_edges},
      {"09_fuse_many_boxes", sample09_fuse_many_boxes},
      {"10_mesh_incremental", sample10_mesh_incremental},
  };
  return kSamples;
}

File: experiments/build123d-vs-ocjs/ocjs/samples.mjs
/**
 * Frontier OCCT workloads mirroring python/samples.py (build123d) and
 * native/samples.cpp.
 *
 * Every sample that batches >1 boolean operation uses the canonical
 * `BRepAlgoAPI_BuilderAlgo` multi-tool form (`SetArguments + SetTools + Build`)
 * — the iterative `Op(prev, next)` chain is the previous-baseline anti-pattern
 * (preserved historically in F13's per-engine 09b/09 + 10b/10 ratios). See
 * F14 — Frontier benchmark sample (see experiments/build123d-vs-ocjs/README.md).
 * Performance — is the canonical comparison and uses these samples directly.
 *
 * Uses `using` declarations (Node 22+/V8 12.4+) for automatic OCCT handle
 * disposal, matching the convention in tests/smoke/*.test.ts.
 */

/** @param {import('../../../build-configs/opencascade_single.js').OpenCascadeInstance} oc */

export function sample01_primitiveBox(oc) {
  using box = new oc.BRepPrimAPI_MakeBox(10, 20, 30);
  using shape = box.Shape();
  if (shape.IsNull()) throw new Error('null shape');
}

export function sample02_primitiveCylinder(oc) {
  using cyl = new oc.BRepPrimAPI_MakeCylinder(5, 15);
  using shape = cyl.Shape();
  if (shape.IsNull()) throw new Error('null shape');
}

export function sample03_booleanFuse(oc) {
  using box1 = new oc.BRepPrimAPI_MakeBox(10, 10, 10);
  using box2 = new oc.BRepPrimAPI_MakeBox(5, 5, 5);
  using s1 = box1.Shape();
  using s2 = box2.Shape();
  using pr = new oc.Message_ProgressRange();
  using fuse = new oc.BRepAlgoAPI_Fuse(s1, s2, pr);
  using pr2 = new oc.Message_ProgressRange();
  fuse.Build(pr2);
  using out = fuse.Shape();
  if (out.IsNull()) throw new Error('null fuse');
}

// Multi-tool BRepAlgoAPI_Cut: 1 BOPDS init over base + 25 cylinder tools,
// instead of 25 separate inits over (current, tool). See F14.
export function sample04_booleanCutGrid(oc) {
  using baseMaker = new oc.BRepPrimAPI_MakeBox(100, 100, 10);
  const tools = [];
  try {
    for (let i = 0; i < 5; i++) {
      for (let j = 0; j < 5; j++) {
        using origin = new oc.gp_Pnt(10 + i * 20, 10 + j * 20, -5);
        using zDir = new oc.gp_Dir(0, 0, 1);
        using ax = new oc.gp_Ax2(origin, zDir);
        using cylMaker = new oc.BRepPrimAPI_MakeCylinder(ax, 2, 20);
        tools.push(cylMaker.Shape());
      }
    }
    using args = new oc.NCollection_List_TopoDS_Shape();
    using toolsList = new oc.NCollection_List_TopoDS_Shape();
    using base = baseMaker.Shape();
    args.Append(base);
    for (const t of tools) toolsList.Append(t);
    using cut = new oc.BRepAlgoAPI_Cut();
    cut.SetArguments(args);
    cut.SetTools(toolsList);
    using pr = new oc.Message_ProgressRange();
    cut.Build(pr);
    using out = cut.Shape();
    if (out.IsNull()) throw new Error('null cut grid');
  } finally {
    for (const t of tools) t.delete();
  }
}

export function sample05_loftThruSections(oc) {
  using loft = new oc.BRepOffsetAPI_ThruSections(true, false, 1e-6);
  const profiles = [
    { z: 0, radius: 10 },
    { z: 15, radius: 5 },
    { z: 30, radius: 8 },
  ];
  for (const { z, radius } of profiles) {
    using center = new oc.gp_Pnt(0, 0, z);
    using dir = new oc.gp_Dir(0, 0, 1);
    using ax = new oc.gp_Ax2(center, dir);
    using circle = new oc.Geom_Circle(ax, radius);
    using edge = new oc.BRepBuilderAPI_MakeEdge(circle);
    using e = edge.Edge();
    using wireMaker = new oc.BRepBuilderAPI_MakeWire(e);
    using w = wireMaker.Wire();
    loft.AddWire(w);
  }
  loft.CheckCompatibility(false);
  using shape = loft.Shape();
  if (shape.IsNull()) throw new Error('null loft');
}

export function sample06_pipeShellSweep(oc) {
  using p1 = new oc.gp_Pnt(0, 0, 0);
  using p2 = new oc.gp_Pnt(0, 0, 30);
  using spineEdge = new oc.BRepBuilderAPI_MakeEdge(p1, p2);
  using eSpine = spineEdge.Edge();
  using spineWire = new oc.BRepBuilderAPI_MakeWire(eSpine);

  using axOrigin = new oc.gp_Pnt(0, 0, 0);
  using axDir = new oc.gp_Dir(0, 0, 1);
  using ax = new oc.gp_Ax2(axOrigin, axDir);
  using circle = new oc.Geom_Circle(ax, 5);
  using profileEdge = new oc.BRepBuilderAPI_MakeEdge(circle);
  using eProf = profileEdge.Edge();
  using profileWire = new oc.BRepBuilderAPI_MakeWire(eProf);

  using sw = spineWire.Wire();
  using pipeShell = new oc.BRepOffsetAPI_MakePipeShell(sw);
  using pw = profileWire.Wire();
  pipeShell.Add(pw, false, false);
  using progress = new oc.Message_ProgressRange();
  pipeShell.Build(progress);
  pipeShell.MakeSolid();
  using shape = pipeShell.Shape();
  if (shape.IsNull()) throw new Error('null pipe');
}

export function sample07_surfaceFillingPatch(oc) {
  using p1 = new oc.gp_Pnt(0, 0, 0);
  using p2 = new oc.gp_Pnt(10, 0, 0);
  using p3 = new oc.gp_Pnt(10, 10, 0);
  using p4 = new oc.gp_Pnt(0, 10, 0);
  using em1 = new oc.BRepBuilderAPI_MakeEdge(p1, p2);
  using em2 = new oc.BRepBuilderAPI_MakeEdge(p2, p3);
  using em3 = new oc.BRepBuilderAPI_MakeEdge(p3, p4);
  using em4 = new oc.BRepBuilderAPI_MakeEdge(p4, p1);
  using e1 = em1.Edge();
  using e2 = em2.Edge();
  using e3 = em3.Edge();
  using e4 = em4.Edge();

  using filling = new oc.BRepOffsetAPI_MakeFilling(
    3, 15, 2, false, 1e-3, 1e-4, 1e-1, 0.1, 8, 9,
  );
  filling.Add(e1, oc.GeomAbs_Shape.GeomAbs_C0, true);
  filling.Add(e2, oc.GeomAbs_Shape.GeomAbs_C0, true);
  filling.Add(e3, oc.GeomAbs_Shape.GeomAbs_C0, true);
  filling.Add(e4, oc.GeomAbs_Shape.GeomAbs_C0, true);
  using pr = new oc.Message_ProgressRange();
  filling.Build(pr);
  using face = filling.Shape();
  if (face.IsNull()) throw new Error('null fill');
}

export function sample08_filletAllEdges(oc) {
  using box = new oc.BRepPrimAPI_MakeBox(20, 20, 20);
  using boxShape = box.Shape();
  using fillet = new oc.BRepFilletAPI_MakeFillet(
    boxShape,
    oc.ChFi3d_FilletShape.ChFi3d_Rational,
  );
  using explorer = new oc.TopExp_Explorer(
    boxShape,
    oc.TopAbs_ShapeEnum.TopAbs_EDGE,
    oc.TopAbs_ShapeEnum.TopAbs_SHAPE,
  );
  while (explorer.More()) {
    using cur = explorer.Current();
    using edge = oc.TopoDS.Edge(cur);
    fillet.Add(3, edge);
    explorer.Next();
  }
  using pr = new oc.Message_ProgressRange();
  fillet.Build(pr);
  using shape = fillet.Shape();
  if (shape.IsNull()) throw new Error('null fillet');
}

// Multi-tool BRepAlgoAPI_Fuse: 1 BOPDS init over args + 39 box tools,
// instead of 39 separate (prev, next) inits in a chain. See F14.
function buildOverlappingBoxesShapes(oc, count = 40, spacing = 3.0, side = 4.0) {
  const shapes = [];
  for (let i = 0; i < count; i++) {
    using origin = new oc.gp_Pnt(i * spacing, 0, 0);
    using maker = new oc.BRepPrimAPI_MakeBox(origin, side, side, side);
    shapes.push(maker.Shape());
  }
  return shapes;
}

function fuseOverlappingBoxesMultiTool(oc, shapes) {
  using args = new oc.NCollection_List_TopoDS_Shape();
  using tools = new oc.NCollection_List_TopoDS_Shape();
  args.Append(shapes[0]);
  for (let i = 1; i < shapes.length; i++) tools.Append(shapes[i]);
  using fuse = new oc.BRepAlgoAPI_Fuse();
  fuse.SetArguments(args);
  fuse.SetTools(tools);
  using pr = new oc.Message_ProgressRange();
  fuse.Build(pr);
  return fuse.Shape();
}

export function sample09_fuseManyBoxes(oc) {
  const shapes = buildOverlappingBoxesShapes(oc);
  let outShape = null;
  try {
    outShape = fuseOverlappingBoxesMultiTool(oc, shapes);
    if (outShape.IsNull()) throw new Error('null multi-tool fuse');
  } finally {
    for (const s of shapes) s.delete();
    if (outShape) outShape.delete();
  }
}

export function sample10_meshIncremental(oc) {
  const shapes = buildOverlappingBoxesShapes(oc);
  let outShape = null;
  try {
    outShape = fuseOverlappingBoxesMultiTool(oc, shapes);
    if (outShape.IsNull()) throw new Error('null multi-tool fuse');
    using mesh = new oc.BRepMesh_IncrementalMesh(outShape, 0.25, false, 0.5, false);
    void mesh;
    using loc = new oc.TopLoc_Location();
    const meshPurposeNone = 0;
    using expf = new oc.TopExp_Explorer(
      outShape,
      oc.TopAbs_ShapeEnum.TopAbs_FACE,
      oc.TopAbs_ShapeEnum.TopAbs_SHAPE,
    );
    if (!expf.More()) throw new Error('no faces');
    using fc = expf.Current();
    using face = oc.TopoDS.Face(fc);
    using tri = oc.BRep_Tool.Triangulation(face, loc, meshPurposeNone);
    if (tri.NbTriangles() < 1) throw new Error('no triangles');
  } finally {
    for (const s of shapes) s.delete();
    if (outShape) outShape.delete();
  }
}

export const SAMPLES = {
  '01_primitive_box': sample01_primitiveBox,
  '02_primitive_cylinder': sample02_primitiveCylinder,
  '03_boolean_fuse': sample03_booleanFuse,
  '04_boolean_cut_grid': sample04_booleanCutGrid,
  '05_loft_thru_sections': sample05_loftThruSections,
  '06_pipe_shell_sweep': sample06_pipeShellSweep,
  '07_surface_filling_patch': sample07_surfaceFillingPatch,
  '08_fillet_all_edges': sample08_filletAllEdges,
  '09_fuse_many_boxes': sample09_fuseManyBoxes,
  '10_mesh_incremental': sample10_meshIncremental,
};
